lib/streamlit/elements/vega_charts.py

```diff
@@ -185,12 +204,70 @@ def id_transform(data) -> dict[str, str]:
         with alt.data_transformers.enable("id"):  # type: ignore[attr-defined,unused-ignore]
             chart_dict = altair_chart.to_dict()
 
-    # Put datasets back into the chart dict but note how they weren't
-    # transformed.
+    # Put datasets back into the chart dict:
     chart_dict["datasets"] = datasets
     return chart_dict
 
 
+def _reset_counter_pattern(prefix: str, vega_spec: str) -> str:
+    """Altair uses a global counter for unnamed parameters and views.
+    We need to reset these counters on a spec-level to make the
+    spec stable across reruns and avoid changes to the element ID.
+    """
+    pattern = re.compile(rf'"{prefix}\d+"')
+    # Get all matches without duplicates in order of appearance.
+    # Using a set here would not guarantee the order of appearance,
+    # which might lead to different replacements on each run.
+    if matches := list(dict.fromkeys(pattern.findall(vega_spec))):
```

Comment:
will the order of keys in the spec will always be the same (is altair guaranteeing that)?

Reply:
I'm pretty certain that there isn't any randomness on Altair side when it comes to how the spec is generated. It might change with Altair updates, but that's not really a case that is relevant for us since we mainly care about having this stable within a session.

Reply:
great! it could make sense to capture this thought also as a comment here

Reply:
Added more comments